use correct reference when using strip_component

b4ldr opened the pull request:
The current implmnetation sets the environment name to the "striped"
version of the name, this is then later used as the git reference when
syncing environments.

With the following configuration
```yaml
sources:
  b4ldr:
    strip_component: "wmcs/"
    ignore_branch_prefixes:
    - production
    - ops
    - sandbox
    - pontoon
    remote: https://github.com/b4ldr/puppet-1
```

where upstream has a branch with wmcs/production then we see the
following error

  Unable to sync repo to unresolvable ref 'production' at
    /etc/puppet/code/environments/production/.git

If upstream also has a production branch then the above configuration
ends up syncing  /etc/puppet/code/environments/production/ with the
remote production branch.  even thought it should be striped via
`ignore_branch_prefixes`

This PR updates the name object to keep hold of the original name and
store it as the ref

b4ldr committed:
use correct reference when using strip_component

The current implmnetation sets the environment name to the "striped"
version of the name, this is then later used as the git reference when
syncing environments.

With the following configuration
```yaml
sources:
  b4ldr:
    strip_component: "wmcs/"
    ignore_branch_prefixes:
    - production
    - ops
    - sandbox
    - pontoon
    remote: https://github.com/b4ldr/puppet-1
```

where upstream has a branch with wmcs/production then we see the
following error

  Unable to sync repo to unresolvable ref 'production' at
    /etc/puppet/code/environments/production/.git

If upstream also has a production branch then the above configuration
ends up syncing  /etc/puppet/code/environments/production/ with the
remote production branch.  even thought it should be striped via
`ignore_branch_prefixes`

This PR updates the name object to keep hold of the original name and
store it as the ref

Magisus reviewed (approved):
Other than that small but important update to the commit message, I think this looks good! It definitely fixes the issue.

reidmv commented on lib/r10k/environment/name.rb:
```diff
@@ -7,7 +7,7 @@ class Name
       # @!attribute [r] name
       #   @return [String] The unmodified name of the environment
-      attr_reader :name
+      attr_reader :name, :ref
```
Possibly a nit: this new attribute, `:ref`, is not documented. I think it needs to be. Without understanding exactly what it is for, and what the difference between it and name is, it would probably be a little bit confusing to look at later and try to figure out what was going on and why.

b4ldr replied:
have added something now

mwaggett commented on lib/r10k/environment/name.rb:
```diff
@@ -5,9 +5,12 @@ module Environment
-      attr_reader :name
+      #   @return [String] The name of the environment (with component striped)
+      attr_reader :name, :ref
+      attr_reader :name, :ref
```
This doesn't need to be here twice.

mwaggett commented on lib/r10k/environment/name.rb:
```diff
@@ -5,9 +5,12 @@ module Environment
       # @!attribute [r] name
-      #   @return [String] The unmodified name of the environment
-      attr_reader :name
+      #   @return [String] The name of the environment (with component striped)
```
What does 'component' refer to?

mwaggett reviewed (commented):
Would you mind adding an entry to the 'Unreleased' section of the CHANGELOG describing this fix?

mwaggett commented on spec/unit/environment/name_spec.rb:
```diff
@@ -149,6 +165,8 @@
         it "replaces invalid characters in #{branch} with underscores" do
           bn = described_class.new(branch.dup, {:correct => true})
           expect(bn.dirname).to eq branch.gsub(/\W/, '_')
+          expect(bn.name).to eq branch
```
huh, I would have thought this would also get corrected, but I guess I'm not totally sure where all the `name` is used.

reidmv replied:
Yeah. After taking a look through it, I realized that it's currently inconsistent. The original implementation of `strip_component` introduced the inconsistency. I'll take responsibility for that. There would ideally be a clear separation between the input value given to `R10K::Environment::Name` and the derived code-environment name on disk (currently called `#dirname`). Bit of work to clean it all up at this point.

mwaggett replied:
ahh. Do you think it's a worthwhile bit of work to do eventually? If so, would you mind opening a ticket for it?

reidmv committed:
Generalize new R10K::Environment::Name attribute

The utility value is in being able to differentiate between the string
used to create the R10K::Environment::Name object, and the #name
attribute return value, which may have been transformed or modified from
the original input.

In order not to build in assumptions about where the original name came
from (it isn't necessarily related to Git), use the general method
`original_name` to differentiate from `name` (rather than the
Git-specific term `ref`).

This commit also changes some method names in R10K::Source::Git to
clarify there the difference between a branch name and an environment
name.

reidmv committed:
Fix discovered rspec errors

reidmv reviewed (approved):
I have a commit in here myself now so somebody should sign off on that, but I'm 👍  overall

Magisus reviewed (approved):
Thanks for weighing in here, Reid, I agree, this is better.

Magisus committed:
(maint) Update Changelog

b4ldr committed:
Merge branch 'main' into fix_strip_component

b4ldr committed:
Update CHANGELOG.mkd
